internal/decode/decode.go

```diff
@@ -0,0 +1,228 @@
+// different shapes. For example,
+//
+// 	// Allow the user to just use the string "default" for the default
+// 	// configuration.
+// 	var name string
+// 	if err := into(&name); err == nil {
+// 		if name == "default" {
+// 			*self = DefaultConfiguration
+// 			return
+// 		}
+// 		return fmt.Errorf("unknown name %q", name)
+// 	}
+//
+// 	// Otherwise, the user must provide {someAttr: "value"} as the input for
+// 	// explicit configuration.
+// 	var custom struct{ SomeAttr string }
+// 	if err := into(&custom); err != nil {
+// 		return err
+// 	}
+//
+// 	self.SomeAttr = custom
+// 	return nil
+//
+// If the destination type or any sub-type implements Decoder, that function
+// will be called. This means that Into MUST NOT be called on the type whose
+// Decode function is currently running or this will end up in an infinite
+// loop.
+type Into func(dest interface{}) error
+
+// decodeFrom builds a decode Into function that reads the given value into
+// the destination.
+func decodeFrom(src interface{}) Into {
+	return func(dest interface{}) error {
+		cfg := mapstructure.DecoderConfig{
+			ErrorUnused: true,
+			Result:      dest,
+			DecodeHook: mapstructure.ComposeDecodeHookFunc(
+				mapstructure.StringToTimeDurationHookFunc(),
+				decoderDecodeHook,
+			),
+			TagName: _tagName,
+		}
+
+		decoder, err := mapstructure.NewDecoder(&cfg)
+		if err != nil {
+			return fmt.Errorf("failed to set up decoder: %v", err)
+		}
+
+		return decoder.Decode(src)
+	}
+}
+
+// decoderDecodeHook is a DecodeHook for mapstructure which recognizes types
+// that implement the Decoder interface.
+func decoderDecodeHook(from, to reflect.Type, data interface{}) (interface{}, error) {
+	if data == nil {
+		return data, nil
+	}
+	out, err := _decoderDecodeHook(from, to, reflect.ValueOf(data))
+	return out.Interface(), err
+}
+
+func _decoderDecodeHook(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+	// Get rid of pointers in either direction. This lets us parse **foo into
+	// a foo where *foo implements Decoder, for example.
+	switch {
+	case from == to:
+		return data, nil
+	case from.Kind() == reflect.Ptr: // *foo => foo
+		return _decoderDecodeHook(from.Elem(), to, data.Elem())
+	case to.Kind() == reflect.Ptr: // foo => *foo
+		out, err := _decoderDecodeHook(from, to.Elem(), data)
+		if err != nil {
+			return out, err
+		}
+
+		// If we didn't know what to do with the input, the returned value
+		// will just be the data as-is and it won't have the correct type.
+		if !out.Type().AssignableTo(to.Elem()) {
+			return data, nil
+		}
+
+		result := reflect.New(to.Elem())
+		result.Elem().Set(out)
+		return result, nil
+	}
+
+	// After eliminating pointers, only destinations whose pointers implement
+	// Decoder are supported. Everything else gets the value unchanged.
+
+	if !reflect.PtrTo(to).Implements(_typeOfDecoder) {
+		return data, nil
+	}
+
+	// value := new(foo)
+	// err := value.Decode(...)
+	// return *value, err
```

Comment:
No, it's just documentation for reviewers to explain what the reflection is doing